main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java

```diff
@@ -379,6 +400,114 @@ private void loadFeedInfo(GtfsTableContainer<GtfsFeedInfo> feedTable) {
     }
   }
 
+  /**
+   * Loads the service date range by determining the earliest start date and the latest end date for
+   * all services referenced with a trip\_id in `trips.txt`. It handles three cases: 1. When only
+   * `calendars.txt` is used. 2. When only `calendar\_dates.txt` is used. 3. When both
+   * `calendars.txt` and `calendar\_dates.txt` are used.
+   *
+   * @param tripContainer the container for `trips.txt` data
+   * @param calendarTable the container for `calendars.txt` data
+   * @param calendarDateTable the container for `calendar\_dates.txt` data
+   */
+  public void loadServiceDateRange(
+      GtfsTableContainer<GtfsTrip> tripContainer,
+      GtfsTableContainer<GtfsCalendar> calendarTable,
+      GtfsTableContainer<GtfsCalendarDate> calendarDateTable) {
+    List<GtfsTrip> trips = tripContainer.getEntities();
+
+    LocalDate earliestStartDate = null;
+    LocalDate latestEndDate = null;
+    if ((calendarDateTable == null) && (calendarTable != null)) {
+      // When only calendars.txt is used
+      List<GtfsCalendar> calendars = calendarTable.getEntities();
+      for (GtfsTrip trip : trips) {
+        String serviceId = trip.serviceId();
+        for (GtfsCalendar calendar : calendars) {
+          if (calendar.serviceId().equals(serviceId)) {
+            LocalDate startDate = calendar.startDate().getLocalDate();
+            LocalDate endDate = calendar.endDate().getLocalDate();
+
+            if (startDate != null || endDate != null) {
+              if (earliestStartDate == null || startDate.isBefore(earliestStartDate)) {
+                earliestStartDate = startDate;
+              }
+              if (latestEndDate == null || endDate.isAfter(latestEndDate)) {
+                latestEndDate = endDate;
+              }
+            }
+          }
+        }
+      }
+    } else if ((calendarDateTable != null) && (calendarTable == null)) {
+      // When only calendar_dates.txt is used
+      List<GtfsCalendarDate> calendarDates = calendarDateTable.getEntities();
+      for (GtfsTrip trip : trips) {
+        String serviceId = trip.serviceId();
+        for (GtfsCalendarDate calendarDate : calendarDates) {
```

Comment:
Can't this loop be replaced by something like this:
GtfsCalendarDate calendarDate = ((GtfsCalendarDateTableContainer) calendarDateTable).byServiceId(serviceId);

Reply:
Same here.